netbox_onboarding/netbox_keeper.py

```diff
@@ -0,0 +1,373 @@
+        self.nb_platform = None
+
+        self.device = None
+        self.nb_mgmt_ifname = None
+        self.nb_primary_ip = None
+
+    def ensure_device_site(self):
+        try:
+            self.nb_site = Site.objects.get(slug=self.netdev_nb_site_slug)
+        except Site.DoesNotExist:
+            raise OnboardException(
+                reason="fail-config",
+                message=f"Site not found: {self.netdev_nb_site_slug}"
+            )
+
+    def ensure_device_manufacturer(self,
+                                   create_manufacturer=PLUGIN_SETTINGS["create_manufacturer_if_missing"],
+                                   ):
+        # First ensure that the vendor, as extracted from the network device exists
+        # in NetBox.  We need the ID for this vendor when ensuring the DeviceType
+        # instance.
+
+        nb_manufacturer_slug = slugify(self.netdev_vendor)
+
+        try:
+            self.nb_manufacturer = Manufacturer.objects.get(
+                slug=nb_manufacturer_slug
+            )
+        except Manufacturer.DoesNotExist:
+            if create_manufacturer:
+                self.nb_manufacturer = Manufacturer.objects.create(
+                    name=self.netdev_vendor,
+                    slug=nb_manufacturer_slug
+                )
+            else:
+                raise OnboardException(
+                    reason="fail-config",
+                    message=f"ERROR manufacturer not found: {self.netdev_vendor}"
+                )
+
+    def ensure_device_type(
+            self,
+            create_device_type=PLUGIN_SETTINGS["create_device_type_if_missing"],
+    ):
+        """Ensure the Device Type (slug) exists in NetBox associated to the netdev "model" and "vendor" (manufacturer).
+
+        Args:
+          #create_manufacturer (bool) :Flag to indicate if we need to create the manufacturer, if not already present
+          create_device_type (bool): Flag to indicate if we need to create the device_type, if not already present
+        Raises:
+          OnboardException('fail-config'):
+            When the device vendor value does not exist as a Manufacturer in
+            NetBox.
+
+          OnboardException('fail-config'):
+            When the device-type exists by slug, but is assigned to a different
+            manufacturer.  This should *not* happen, but guard-rail checking
+            regardless in case two vendors have the same model name.
+        """
+        # Now see if the device type (slug) already exists,
+        #  if so check to make sure that it is not assigned as a different manufacturer
+        # if it doesn't exist, create it if the flag 'create_device_type_if_missing' is defined
+
+        slug = self.netdev_model
+        if re.search(r"[^a-zA-Z0-9\-_]+", slug):
+            logging.warning("device model is not sluggable: %s", slug)
+            self.netdev_model = slug.replace(" ", "-")
+            logging.warning("device model is now: %s", self.netdev_model)
```

Comment:
Can/should we use `slugify` here instead of this whole block?

Reply:
I intentionally excluded this from this PR, I will create another issue to track this